src/Nethermind.Verkle/Proofs/MultiProof.cs

```diff
@@ -118,20 +118,38 @@ public bool CheckMultiProof(Transcript transcript, VerkleVerifierQuery[] queries
         transcript.AppendPoint(d, "D");
         FrE t = transcript.ChallengeScalar("t");
 
-        FrE[] g2Den = new FrE[queries.Length];
+        FrE[] groupedEvals = new FrE[domainSize];
+        // Calculate groupedEvals = r * y_i.
         for (int i = 0; i < queries.Length; i++)
         {
-            g2Den[i] = t - FrE.SetElement(u0: queries[i].ChildIndex);
+            groupedEvals[queries[i].ChildIndex] += powersOfR[i] * queries[i].ChildHash;
         }
-        g2Den = FrE.MultiInverse(g2Den);
+
+
+        // REMOVABLE COMMENT: The idea here is to avoid doing a lot of repeated inverses. If we have thousands of 
+        // queries to check, there can only be up to 256 evaluations points.
+        FrE[] helperScalarDens = new FrE[domainSize];
+        // Compute helperScalarsDen = 1 / (t - z_i).
+        for (int i = 0; i < domainSize; i++)
+        {
+            helperScalarDens[i] = t - FrE.SetElement(u0: queries[i].ChildIndex);
+        }
+        helperScalarDens = FrE.MultiInverse(helperScalarDens);
 
         FrE g2T = FrE.Zero;
+        for (int i = 0; i < domainSize; i++)
+        {
+            // NIT: could the == operation be defined for FrEs? :)
+            if (groupedEvals[i].Equals(FrE.Zero)) continue;
```

Comment:
Actually, we can remove this line if the multiplication below is fast for 0 elements. Up to you!